### PR TITLE
[luci] Add a method to handle all positive & negative values in computing asymmetric scale

### PR DESCRIPTION
**Pull request by meejeong**
- The previous lines(```L74-L78```) were meaningful only in the case of ```symmetric quantization```.
- For handling all positive & negative values, I used different scaling factor
- When the minimum and maximum values ​​are the same, the processing method changed.
  (It related to all positive & negative values, So I uploaded the same PR.)
image

ONE-DCO-1.0-Signed-off-by: mia park <mia.park@samsung.com>

Related to: #696, https://github.sec.samsung.net/STAR/nnapps/issues/379

### CODE REVIEW

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 1]
```diff
@@ -61,31 +61,38 @@ void compute_sym_scale_zp(float min, float max, float &scaling_factor, int64_t &
 void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t &zp,
                            float &nudged_min, float &nudged_max)
 {
-  assert(min != max);
 
+  LOGGER(l);
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:65]: Please swap empty line 64 and 65

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 2]
```diff
@@ -62,31 +62,38 @@ void compute_sym_scale_zp(float min, float max, float &scaling_factor, int64_t &
 void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t &zp,
                            float &nudged_min, float &nudged_max)
 {
-  assert(min != max);
 
+  LOGGER(l);
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:66]: same here

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 3]
```diff
@@ -61,31 +61,38 @@ void compute_sym_scale_zp(float min, float max, float &scaling_factor, int64_t &
...
   double scale = (rmax - rmin) / (qmax_double - qmin_double);
-  const double zero_point_from_min = qmin_double - rmin / scale;
-  const double zero_point_from_max = qmax_double - rmax / scale;
-  const double zero_point_from_min_error = std::abs(qmin_double) + std::abs(rmin / scale);
-  const double zero_point_from_max_error = std::abs(qmax_double) + std::abs(rmax / scale);
-  const double zero_point_double = zero_point_from_min_error < zero_point_from_max_error
-                                       ? zero_point_from_min
-                                       : zero_point_from_max;
-  uint8_t nudged_zero_point = 0;
+  double zero_point_double{0};
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:73]: above uses style `int32_t kMinScale = 0` and this uses `zero_point_double{0}`. it would be better to use one style, for here ` = 0` looks preferable.

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 4]
```diff
@@ -61,31 +61,38 @@ void compute_sym_scale_zp(float min, float max, float &scaling_factor, int64_t &
...
+  {
+    WARN(l) << "The minimum and maximum values are the same." << std::endl;
+    if (min >= 0 && max >= 0)
+      zero_point_double = kMinScale;
+    else
+      zero_point_double = kMaxScale;
+  }
+  else
+    zero_point_double = qmin_double - rmin / scale;
+  uint8_t nudged_zero_point{0};
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:84]: this line would be better to go after `double zero_point_double{0}`

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 5]
```diff
@@ -61,31 +61,38 @@ void compute_sym_scale_zp(float min, float max, float &scaling_factor, int64_t &
...
   if (zero_point_double <= qmin_double)
   {
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:86]: Can you add the assertion for each case?

`assert(min >= 0 && max >= 0);`
`assert(min < 0 && max < 0);`
`assert(min < 0 && max >= 0);`

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 6]
```diff
@@ -61,31 +61,38 @@ void compute_sym_scale_zp(float min, float max, float &scaling_factor, int64_t &
 void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t &zp,
                            float &nudged_min, float &nudged_max)
 {
-  assert(min != max);
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:64]: Can you add an assertion if you don't handle the case min <= max.

`assert(min <= max);`

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 7]
```diff
@@ -61,34 +61,45 @@ void compute_sym_scale_zp(float min, float max, float &scaling_factor, int64_t &
...
+  LOGGER(l);
 
+  assert(min <= max);
   const int32_t kMinScale = 0;
   const int32_t kMaxScale = 255;
   const double qmin_double = kMinScale;
   const double qmax_double = kMaxScale;
   const double rmin = std::fmin(0, min);
   const double rmax = std::fmax(0, max);
-
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:72]: Please split this change to another PR, maybe you can call it enhance readability.

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 8]
```diff
@@ -62,34 +62,45 @@ void compute_sym_scale_zp(float min, float max, float &scaling_factor, int64_t &
...
 
+  assert(min <= max);
   const int32_t kMinScale = 0;
   const int32_t kMaxScale = 255;
   const double qmin_double = kMinScale;
   const double qmax_double = kMaxScale;
   const double rmin = std::fmin(0, min);
   const double rmax = std::fmax(0, max);
-
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:73]: ditto

**Comment by seanshpark** [comment]
I didn't notice the changes; I didn't see any pings from the system -_-;;;
Sorry for late review...
